subprojects/base-services/src/main/java/org/gradle/internal/reflect/DirectInstantiator.java

```diff
@@ -36,7 +39,14 @@ private DirectInstantiator() {
 
     public <T> T newInstance(Class<? extends T> type, Object... params) {
         try {
-            Constructor<?> match = doGetConstructor(type, constructorCache.get(type), params);
+            Class<?>[] argTypes = wrapArgs(params);
+            Constructor<?> match = null;
+            while (match == null) {
```

Comment:
I played with this, but the problem is that we need 2 things:

1. the section of code that uses the cache value then needs to be wrapped into a block of code (`Action<T>` or something similar), which involves usage of capture variables to be able to return something from this (for example, we would need to construct inside the `Action<T>`, assign to a single entry of a final array so that we can return the value)
2. We need, to be safe, to have `Action<Wrapper<T>>` where the wrapper would store the value, so that it is strongly referenced, and `T`, so that we can have access to the extra data.

I realized it wasn't worth the effort.